### ISSUE TITLE
Deploying a new version of a WAP via kudu always touches web.config

### ISSUE THREAD

**Issue by davidfowl** [projectkudu/kudu#1167]
Seems like this is the code responsible:

https://github.com/projectkudu/kudu/blob/718d721c5a8949b574f4077b3eeb4a0cf6ae284a/Kudu.Core/Deployment/DeploymentManager.cs#L606

This is pretty annoying as tiny view changes cause an appdomain recycle.

**Comment by onashackem** [comment]
+1! I was going to create a new issue but I see there is already one.

**Comment by jirihelmich** [comment]
Affecting our project as well. +1

**Comment by davidebbo** [comment]
So I'm wondering if this issue is strictly about web.config or if there is more to it. If you has an ASP.NET WAP, then msbuild needs to build a DLL that goes in bin from your sources. The wait deployment works in Kudu is:

1. Build the solution into a clean temporary folder
2. Use kudusync to copy the relevant files to wwwroot

Problem is that the DLL built in step 1 will always be different, and always end up getting copied to bin. This will cause an AppDomain restart regardless of whether web.config is touched.

**Comment by davidebbo** [comment]
Actually, after actually trying it, I see that my previous comment is not quite correct. As part of step 1, binaries first get built in-place in the repository folder (e.g. in `obj\Release`), and then copied to the temp folder. So if no actually code change happens, `obj\Release` is not rebuilt, and the temp folder ends up with exactly the same files as last time. So the kudusync optimization kicks in, and they don't get copied.

So the web.config optimization would make sense.

Since doing it automatically is tricky, how would people feel about a manual App Setting for this? From my earlier comment:

> Or if all else fails, we could come up with a new flag, e.g. SCM_TOUCH_CONFIG_ON_DEPLOYMENT. It would be true by default to match current behavior, but people could override if they want.

It would be cheap to implement, and would be much less likely to cause regressions.

**Comment by siroky** [comment]
There are two parts of this issue. The first one, which we consider pretty serious bug, is caused by the following sequence of events:

1. During Kudusync, the `Web.config` is copied first and the subfolders including the `bin` are copied after that. As I desribed earlier, as soon as the `Web.config` changes the web application is immediately restarted (see http://beweb.pbworks.com/w/page/30073098/Prevent%20app%20restarts%20and%20site%20downtime%20when%20deploying%20files) with the `HostingEnvironment.ShutdownReason` set to `ConfigurationChange`.
2. New instance of the application is started, but **not all dlls might have been copied yet**, since it happens in the middle of Kudusync. This depends on how much Kudusync has to copy and how is the app being utilized at the moment. The new application can start with half of the new dlls (already copied) and half of the old dlls (not copied yet). I can imagine this inconsistent state may lead to big problems in some use cases, e.g. if the app performs some database migrations during startup.
3. After all files are copied and Kudusync finished, the app is restarted again, now with shutdown reason `BinDirChangeOrDirectoryRename`.

The solution for this (also suggested by the article I linked) is to simply **copy the `Web.config` as the last file** during Kudusync. As a consequnce, the redeployment time during which the app is not responsive will become less than half of what it's currently (two restarts and some idle time between them will be reduced to just one restart).

The second part of this issue is the proposed optimization not to touch web.config if it doesn't change. This is just a "nice to have" for us since it doesn't solve the first part, which is much more critical for us. Also note that many apps use Web.Release.config which causes the Web.config to be touched during every build. So the optimization would have to actually compare contents of the old and new web.config and copy the new one only if the content differs from the old one. Comparing file modification times isn't sufficient. Turning this optimization by app settings is ok for me and makes perfect sense since it can change behavior of some deployments.

**Comment by davidebbo** [comment]
Digging into things a bit more, I'm realizing that there are two very distinct code paths that end up touching web.config, and that are I think getting a bit mixed up in this thread:

1. The first is what @siroky describes: web.config always changes because of the Web.Release.config transform, and KuduSync ends up copying, and doing that before other files.
2. The second is what @davidfowl started this issue with: after the whole build/deployment script is complete, we make an explicit call to `TryTouchWebConfig` ([here](https://github.com/projectkudu/kudu/blob/a1b175360318b5d84b13d08082e9684d873d7a08/Kudu.Core/Deployment/DeploymentManager.cs#L610)). This is done for correctness in the Node-like scenarios.

Note that when I was discussing an App Setting earlier, I was referring to using it for 2., and not for 1. And I still think this is the way to go for 2.

But now I see that in addition to that, there is work to do for 1. Though it should be noted that just copying web.config first is not sufficient to solve the original problem, which was about avoiding all restarts when only a trivial change is made.

So to summarize, three potential changes need to be made:

1. avoid copying web.config during KuduSync is the content has not changed.
2. when copying web.config using KuduSync, do it last
3. provide an opt-out for the `TryTouchWebConfig` call

**Comment by ahmelsayed** [comment]
1 and 2 should be fixed by https://github.com/projectkudu/KuduSync.NET/commit/337e5ea24621310b1ea209e424dbed09f83cf5da 
The fix for 3 should go in kudu. I'll work on that now.

**Comment by davidebbo** [comment]
No that fix is not yet deployed to Azure, but should be in a few days.

**Comment by davidebbo** [comment]
And the App Setting to set for 3 is: `SCM_TOUCH_WEBCONFIG_AFTER_DEPLOYMENT=0`

**Comment by siroky** [comment]
That's great news. Thanks!

**Comment by davidebbo** [comment]
The fix is now deployed. Would be great if you could validate your scenarios. Thanks!

**Comment by siroky** [comment]
I did a few test deployments and it seems to be working well. I'll let you know if I encountered something strange on our production environment with multiple instances and regions. 

Just one suggestion, it'd be good to log more information about what's happening. E.g. "Web.config didn't change, not copying." or "Copying Web.config", "Touching Web.config" or "Not touching Web.config" etc). So that other people can deduce from the deployment log what's happening and easily find relevant information in your docs (the `SCM_TOUCH_WEBCONFIG_AFTER_DEPLOYMENT`). Otherwise it's pretty complicated to understand what's actually happening and why is the ASP.NET app being restarted.
